cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -111,29 +122,75 @@
    */
   ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile,
                                   ClassLoader parentClassLoader) throws IOException, InvalidArtifactException {
+    Path tmpDir = Paths.get(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                            cConf.get(Constants.AppFabric.TEMP_DIR)).toAbsolutePath();
+    Files.createDirectories(tmpDir);
+    Location artifactLocation = Locations.toLocation(artifactFile);
 
-    ArtifactClasses.Builder builder = inspectApplications(artifactId, ArtifactClasses.builder(),
-                                                          Locations.toLocation(artifactFile));
-    File tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
-                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
-    File stageDir = DirUtils.createTempDir(tmpDir);
-    try (PluginInstantiator pluginInstantiator = new PluginInstantiator(cConf, parentClassLoader, stageDir)) {
-      pluginInstantiator.addArtifact(Locations.toLocation(artifactFile), artifactId.toArtifactId());
-      inspectPlugins(builder, artifactFile, artifactId.toArtifactId(), pluginInstantiator);
+    Path stageDir = Files.createTempDirectory(tmpDir, artifactFile.getName());
+    try {
+      File unpackedDir = BundleJarUtil.unJar(artifactLocation,
+                                             Files.createTempDirectory(stageDir, "unpacked-").toFile());
+
+      ArtifactClasses.Builder builder = inspectApplications(artifactId, ArtifactClasses.builder(),
+                                                            artifactLocation, unpackedDir);
+
+      try (PluginInstantiator pluginInstantiator =
+             new PluginInstantiator(cConf, parentClassLoader,
+                                    Files.createTempDirectory(stageDir, "plugins-").toFile())) {
+        pluginInstantiator.addArtifact(artifactLocation, artifactId.toArtifactId());
+        inspectPlugins(builder, artifactFile, artifactId.toArtifactId(), pluginInstantiator);
+      }
+      return builder.build();
     } finally {
       try {
-        DirUtils.deleteDirectoryContents(stageDir);
+        DirUtils.deleteDirectoryContents(stageDir.toFile());
       } catch (IOException e) {
         LOG.warn("Exception raised while deleting directory {}", stageDir, e);
       }
     }
-    return builder.build();
+  }
+
+  /**
+   * Extracts all dataset classes defined under the given artifact directory. It will scan every .class files under
+   * the artifact directory as well as .class files inside .jar files in the artifact directory.
+   * This method uses bytecode inspection, hence the actual class is not loaded. This is because loading every single
```

Comment:
yes and it includes non static ones as well. This is because every inner class in Java has it's own separate `.class` file accessible through the classpath.